File: src/239. Sliding Window Maximum/239.py
from typing import List


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        if not nums:
            return []
        win, res = [], []
        for i, num in enumerate(nums):
            if i >= k and win[0] == i - k:
                win.pop(0)
            while win and nums[win[-1]] <= num:
                win.pop()
            win.append(i)
            if i >= k - 1:
                res.append(nums[win[0]])
        return res


if __name__ == "__main__":
    nums, k = [1, 3, -1, -3, 5, 3, 6, 7], 3
    print(Solution().maxSlidingWindow(nums, k))

File: src/239. Sliding Window Maximum/239.go
func maxSlidingWindow(nums []int, k int) []int {
    n := len(nums)

	if n == 0 {
		return make([]int,0)
	}


	i,j := 0,0

	max := nums[i]

	r := make([]int,0)

	pos := 0

	for j < n {

		if nums[j] > max {
			max = nums[j]
			if i + k == j + 1 {
				r = append(r,max)
			}
			pos = j
		} else if i + k == j + 1 {
			r = append(r,max)
		}

		j++

		if (j - i) == k {

			if pos == i && j < n {
				pos++
				max = nums[pos]
				m := pos
				for m < j {
					if nums[m] > max {
						max = nums[m]
						pos = m
					}
                    m++
				}
			}

			i++
		}

	}

	return r
}